crates/circuit/src/dag_circuit.rs

```diff
@@ -2771,43 +2773,121 @@ def _format(operand):
+        let dags = PyList::empty_bound(py);
+
+        for comp_nodes in connected_components.iter() {
+            let mut new_dag = self.copy_empty_like(py)?;
+            new_dag.set_global_phase(py, &PyFloat::new_bound(py, 0 as c_double))?;
+
+            // A map from nodes in the this DAGCircuit to nodes in the new dag. Used for adding edges
+            let mut node_map: HashMap<NodeIndex, NodeIndex> =
+                HashMap::with_capacity(comp_nodes.len());
+
+            // Adding the nodes to the new dag
+            let mut non_classical = false;
+            for node in comp_nodes {
+                match self.dag.node_weight(*node) {
+                    Some(w) => match w {
+                        NodeType::ClbitIn(b) => {
+                            let clbit_in = new_dag.clbit_input_map.get(b).unwrap();
+                            node_map.insert(*node, *clbit_in);
+                        }
+                        NodeType::ClbitOut(b) => {
+                            let clbit_out = new_dag.clbit_output_map.get(b).unwrap();
+                            node_map.insert(*node, *clbit_out);
+                        }
+                        NodeType::QubitIn(q) => {
+                            let qbit_in = new_dag.qubit_input_map.get(q).unwrap();
+                            node_map.insert(*node, *qbit_in);
+                            non_classical = true;
+                        }
+                        NodeType::QubitOut(q) => {
+                            let qbit_out = new_dag.qubit_output_map.get(q).unwrap();
+                            node_map.insert(*node, *qbit_out);
+                            non_classical = true;
+                        }
+                        NodeType::Operation(pi) => {
+                            let qubits = self.qargs_cache.intern(pi.qubits_id);
+                            let clbits = self.cargs_cache.intern(pi.clbits_id);
+                            let qubits_id =
+                                Interner::intern(&mut new_dag.qargs_cache, qubits.clone())?;
+                            let clbits_id =
+                                Interner::intern(&mut new_dag.cargs_cache, clbits.clone())?;
+
+                            let extr_attr = pi.extra_attrs.clone().unwrap();
+                            let new_pi = NodeType::Operation(PackedInstruction::new(
+                                pi.op.clone(),
+                                qubits_id,
+                                clbits_id,
+                                pi.params.clone(),
+                                extr_attr.label,
+                                extr_attr.duration,
+                                extr_attr.unit,
+                                extr_attr.condition,
+                                #[cfg(feature = "cache_pygates")]
+                                pi.py_op.borrow().clone(),
+                            ));
```

Comment:
We do assume that if the target dag was created with `copy_empty_like` that the interners will be the same (they are copied from the source dag into the target one during `copy_empty_like`, so you can probably just replace all of this with `pi.clone()` and call it good.